prac_07/project_management.py

```diff
@@ -0,0 +1,178 @@
+        print(headers, file=out_file)
+        for project in projects:
+            print(repr(project), file=out_file)
+
+
+def display_projects(projects):
+    """Display projects grouped by completion status."""
+    completed_projects = []
+    print("Incomplete projects:")
+    for project in projects:
+        print(project) if not project.is_completed() else completed_projects.append(project)
+    print("Completed projects:")
+    for project in completed_projects:
+        print(project)
+
+
+def filter_projects(projects):
+    """Display projects started after specified date."""
+    date_threshold = get_date("Show projects that start after date")  # Should this be outside the function?
+    for project in projects:
+        start_date = datetime.strptime(project.start_date, "%d/%m/%Y").date()
+        if start_date >= date_threshold:
+            print(project)
+
+
+def add_project(projects):
+    """Add project from user input."""
+    print("Let's add a new project")
+    name = input("Name: ")
+    start_date = get_date("Start date").strftime("%d/%m/%Y")
+    priority = get_number("Priority: ")
+    cost_estimate = float(input("Cost estimate: $"))  # not yet error-checked
+    completion = get_number_in_range(101, "Percent complete: ")
+    projects.append(Project(name, start_date, priority, cost_estimate, completion))
+    projects.sort()
+
+
+def get_project_choice(projects):
+    """Get valid project choice from user."""
+    for i, project in enumerate(projects):
+        print(f"{i} {project}")
+    project_choice = projects[
+        get_number_in_range(len(projects), "Project choice: ")]
+    print(project_choice)
+    return project_choice
+
+
+def update_project(project):
+    """Update the priority and/or completion percentage based on user input."""
+    project.priority = get_new_value(project.priority, "priority")
+    new_percentage = get_new_value(project.completion, "completion")
+    while new_percentage < 0 or new_percentage > 100:
+        print("Invalid percentage")
+        new_percentage = get_new_value(project.completion, "completion")
+    project.completion = new_percentage
+
+
+def get_new_value(field, fieldname):
+    """Return new field value from user input (no change if input is empty string)."""
+    value = input(f"New {fieldname}: ")
+    is_valid = False
+    while not is_valid:
+        try:
+            field = int(value)
+            is_valid = True
+        except ValueError:
+            is_valid = not value  # True if input is empty
+    return field
+
+
+def get_date(prompt):
+    """Get valid date from user input."""
+    is_valid_date = False
+    while not is_valid_date:
+        try:
+            date_string = input(f"{prompt} (dd/mm/yy): ")
+            date = datetime.strptime(date_string, "%d/%m/%Y").date()
+            is_valid_date = True
+        except ValueError:
+            print("Invalid date!")
+    return date  # Ignore Pycharm warning (impossible error)
```

Comment:
Sidenote: can turn this error checking to syntax only so it won't show a bunch of these errors. Not sure if this is recommended.